posthog/models/sessions/sql.py

```diff
@@ -0,0 +1,130 @@
+from django.conf import settings
+
+from posthog.clickhouse.table_engines import (
+    Distributed,
+    ReplicationScheme,
+    AggregatingMergeTree,
+)
+
+TABLE_BASE_NAME = "sessions"
+SESSIONS_DATA_TABLE = lambda: f"sharded_{TABLE_BASE_NAME}"
+
+# if updating these column definitions
+# you'll need to update the explicit column definitions in the materialized view creation statement below
+SESSIONS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    -- part of order by so will aggregate correctly
+    session_id VARCHAR,
+    -- part of order by so will aggregate correctly
+    team_id Int64,
+    -- ClickHouse will pick any value of distinct_id for the session
+    -- this is fine since even if the distinct_id changes during a session
+    -- it will still (or should still) map to the same person
+    distinct_id SimpleAggregateFunction(any, String),
+    min_first_timestamp SimpleAggregateFunction(min, DateTime64(6, 'UTC')),
+    max_last_timestamp SimpleAggregateFunction(max, DateTime64(6, 'UTC')),
+
+    urls SimpleAggregateFunction(groupUniqArrayArray, Array(String)),
+    entry_url AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    exit_url AggregateFunction(argMax, String, DateTime64(6, 'UTC')),
+    initial_utm_source AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_campaign AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_medium AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_term AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_content AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_referring_domain AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+
+    event_count SimpleAggregateFunction(sum, Int64),
+    pageview_count SimpleAggregateFunction(sum, Int64),
+) ENGINE = {engine}
+"""
+
+SESSIONS_DATA_TABLE_ENGINE = lambda: AggregatingMergeTree(TABLE_BASE_NAME, replication_scheme=ReplicationScheme.SHARDED)
+
+SESSIONS_TABLE_SQL = lambda: (
+    SESSIONS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(min_first_timestamp)
+    -- order by is used by the aggregating merge tree engine to
+    -- identify candidates to merge, e.g. toDate(min_first_timestamp)
+    -- would mean we would have one row per day per session_id
+    -- if CH could completely merge to match the order by
+    -- it is also used to organise data to make queries faster
+    -- we want the fewest rows possible but also the fastest queries
+    -- since we query by date and not by time
+    -- and order by must be in order of increasing cardinality
+    -- so we order by date first, then team_id, then distinct_id, then session_id
+    -- hopefully, this is a good balance between the two
+    ORDER BY (toDate(min_first_timestamp), team_id, distinct_id, session_id)
```

Comment:
i added distinct id to the order by since i'm  guessing you'll sometimes query for a persons sessions

Reply:
I'm not sure about this - as distinct_id will often change across a session

Reply:
the fun of aggregated tables :)

`distinct_id SimpleAggregateFunction(any, String),`

for the sessions table we have a single distinct id randomly selected from all distinct_ids seen for that session. 

this order by only helps if we'll ever say select X from sessions where distinct_id = Y